File: sdk_v2/js/src/catalog.ts
// `Catalog` instances are NEVER constructed directly by user code — they are returned from
// `FoundryLocalManager.catalog`. The underlying native catalog is owned by the parent manager; the JS wrapper
// holds a reference to keep the manager alive while the catalog is reachable.
//
// The underlying native catalog operations are synchronous; the async surface here is for parity with the C# /
// Python SDKs. `getModel`, `getModelVariant`, and `getLatestVersion` throw when the alias / id is not found.

import type { NativeCatalog, NativeModel } from "./detail/native.js";
import type { IModel } from "./imodel.js";
import { Model, unwrapNativeModel, wrapNativeModel } from "./model.js";
import { type MutableModelInfo, unwrapMutableModelInfo } from "./modelInfo.js";

const internalCtorKey = Symbol("Catalog.internal");
const nativeByCatalog = new WeakMap<Catalog, NativeCatalog>();

export class Catalog {
  readonly #native: NativeCatalog;

  /** @internal — wraps a native catalog handle. Do not call from user code. */
  constructor(token: typeof internalCtorKey, native: NativeCatalog) {
    if (token !== internalCtorKey) {
      throw new TypeError("Catalog is internal — obtain instances via FoundryLocalManager.catalog");
    }
    this.#native = native;
    nativeByCatalog.set(this, native);
  }

  /** Catalog name (e.g. `"AzureFoundryCatalog"`). */
  get name(): string {
    return this.#native.getName();
  }

  /** All models in the catalog. */
  async getModels(): Promise<IModel[]> {
    return wrapAll(this.#native.getModels());
  }

  /** Every individual model variant currently present in the local cache. */
  async getCachedModels(): Promise<IModel[]> {
    return wrapAll(this.#native.getCachedModels());
  }

  /** Models currently loaded into memory. */
  async getLoadedModels(): Promise<IModel[]> {
    return wrapAll(this.#native.getLoadedModels());
  }

  /**
   * Look up a model by alias. Throws when the alias is not present in the catalog. Callers that want a
   * "find or undefined" shape should iterate `getModels()` themselves.
   */
  async getModel(alias: string): Promise<IModel> {
    validateNonEmptyString(alias, "Model alias");
    const n = this.#native.getModel(alias);
    if (n === undefined) {
      throw new Error(`Model with alias '${alias}' not found.`);
    }
    return wrapNativeModel(n);
  }

  /** Look up a specific model variant by its full model id. Throws when not found. */
  async getModelVariant(modelId: string): Promise<IModel> {
    validateNonEmptyString(modelId, "Model ID");
    const n = this.#native.getModelVariant(modelId);
    if (n === undefined) {
      throw new Error(`Model variant with ID '${modelId}' not found.`);
    }
    return wrapNativeModel(n);
  }

  /** Resolve the latest catalog version for the same model name. Throws when not found. */
  async getLatestVersion(model: IModel): Promise<IModel> {
    if (!(model instanceof Model)) {
      throw new TypeError("Catalog.getLatestVersion: expected a Model instance");
    }
    const n = this.#native.getLatestVersion(unwrapNativeModel(model));
    if (n === undefined) {
      throw new Error(`Latest version for model '${model.alias}' not found.`);
    }
    return wrapNativeModel(n);
  }

  /**
   * Get all versions of a model alias, optionally narrowed to a single variant name. `maxVersions`
   * defaults to 50 and acts as a per-variant cap; pass 0 or a negative value for no cap.
   */
  async getModelVersions(modelAlias: string, modelName?: string, maxVersions = 50): Promise<IModel[]> {
    validateNonEmptyString(modelAlias, "Model alias");
    if (modelName !== undefined) {
      validateNativeString(modelName, "Model name");
    }
    // The native parameter is int32_t; N-API's Int32Value() silently coerces
    // fractions, NaN, Infinity, and out-of-range values into a different cap.
    // Reject them here (matching Rust's i32::MAX rejection) so callers get a
    // clear error rather than a surprising truncated result. Negatives/0 are
    // valid ("no cap") but must still fit i32.
    if (!Number.isInteger(maxVersions) || maxVersions < -(2 ** 31) || maxVersions > 2 ** 31 - 1) {
      throw new TypeError("maxVersions must be an integer within the 32-bit range.");
    }
    return wrapAll(await this.#native.getModelVersions(modelAlias, modelName ?? null, maxVersions));
  }

  /**
   * Register existing local model assets. Use this only with `CatalogType.Local`; public catalogs reject mutation.
   * Registration requires `task`. Display name, publisher, validated runtime metadata, modalities, and custom
   * properties are preserved and may receive authoritative defaults. Identity, alias, type, timestamps, context length,
   * and prompt templates are SDK-derived; caller location and internal metadata are ignored. A caller-supplied provider
   * becomes the default load override; an SDK-derived artifact provider leaves `genai_config.json` options intact. The
   * catalog never deletes `modelPath`.
   */
  async registerModel(modelPath: string, modelId: string, metadata: MutableModelInfo): Promise<IModel> {
    validateRegistrationArgs(modelPath, modelId, metadata);
    return wrapNativeModel(await this.#native.registerModel(modelPath, modelId, unwrapMutableModelInfo(metadata)));
  }

  /**
   * Synchronous registration variant with the same metadata ownership rules as `registerModel()`. This performs file
   * I/O and blocks the event loop; prefer `registerModel()`.
   */
  registerModelSync(modelPath: string, modelId: string, metadata: MutableModelInfo): IModel {
    validateRegistrationArgs(modelPath, modelId, metadata);
    return wrapNativeModel(this.#native.registerModelSync(modelPath, modelId, unwrapMutableModelInfo(metadata)));
  }

  /** Unregister a local model without deleting its assets. Accepts either an alias or a full model ID. */
  async unregisterModel(aliasOrModelId: string): Promise<void> {
    validateNonEmptyString(aliasOrModelId, "Alias or model ID");
    await this.#native.unregisterModel(aliasOrModelId);
  }

  /** Synchronous unregistration variant. This performs file I/O and blocks the event loop. */
  unregisterModelSync(aliasOrModelId: string): void {
    validateNonEmptyString(aliasOrModelId, "Alias or model ID");
    this.#native.unregisterModelSync(aliasOrModelId);
  }
}

/** @internal — used by `FoundryLocalManager` to wrap a native catalog. */
export function wrapNativeCatalog(native: NativeCatalog): Catalog {
  return new Catalog(internalCtorKey, native);
}

/** @internal Test-only access for exercising native async lifetime boundaries. */
export function unwrapNativeCatalog(catalog: Catalog): NativeCatalog {
  const native = nativeByCatalog.get(catalog);
  if (native === undefined) {
    throw new TypeError("Expected a Catalog instance");
  }
  return native;
}

function wrapAll(natives: readonly NativeModel[]): IModel[] {
  return natives.map((n) => wrapNativeModel(n));
}

function validateNonEmptyString(value: string, name: string): void {
  if (typeof value !== "string" || value.trim() === "") {
    throw new TypeError(`${name} must be a non-empty string.`);
  }
  validateNativeString(value, name);
}

function validateNativeString(value: string, name: string): void {
  if (value.includes("\0")) {
    throw new TypeError(`${name} must not contain an embedded NUL character.`);
  }
}

function validateRegistrationArgs(modelPath: string, modelId: string, metadata: MutableModelInfo): void {
  validateNonEmptyString(modelPath, "Model path");
  validateNonEmptyString(modelId, "Model ID");
  unwrapMutableModelInfo(metadata);
}

File: sdk_v2/rust/src/catalog.rs
//! Model catalog — discovery and lookup for available models.
//!
//! The native catalog (owned by the [`FoundryLocalManager`](crate::FoundryLocalManager))
//! caches the model list and refreshes itself, so this is a thin async wrapper
//! that preserves the legacy public surface.

use std::sync::Arc;

use crate::detail::api::Api;
use crate::detail::manager::NativeManager;
use crate::detail::model::Model;
use crate::detail::native::NativeCatalog;
use crate::detail::task::spawn_blocking;
use crate::error::{FoundryLocalError, Result};
use crate::types::{CatalogType, ModelInfoBuilder};

/// The model catalog provides discovery and lookup for all available models.
pub struct Catalog {
    native: NativeCatalog,
    name: String,
    catalog_type: CatalogType,
}

impl Catalog {
    pub(crate) fn new(
        api: Arc<Api>,
        ptr: *mut crate::detail::ffi::flCatalog,
        manager: Arc<NativeManager>,
        catalog_type: CatalogType,
    ) -> Result<Self> {
        let native = NativeCatalog::new(api, ptr, manager);
        let name = native.name().unwrap_or_else(|_| "default".into());
        Ok(Self {
            native,
            name,
            catalog_type,
        })
    }

    /// Catalog name as reported by the native core.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// The source type represented by this catalog.
    pub fn catalog_type(&self) -> CatalogType {
        self.catalog_type
    }

    /// Create mutable metadata backed by the same native runtime as this catalog.
    pub fn create_model_info(&self) -> Result<ModelInfoBuilder> {
        ModelInfoBuilder::new(Arc::clone(&self.native.api), self.native.manager())
    }

    /// Refresh the catalog from the native core.
    ///
    /// **No-op.** The native catalog manages its own caching and refresh, so
    /// there is nothing for the SDK to do here. This method is retained only for
    /// API compatibility with the legacy SDK and always returns `Ok(())`.
    pub async fn update_models(&self) -> Result<()> {
        Ok(())
    }

    /// Return all known models keyed by alias.
    pub async fn get_models(&self) -> Result<Vec<Arc<Model>>> {
        let native = self.native.clone();
        spawn_blocking(move || {
            native
                .get_models()?
                .into_iter()
                .map(|m| Model::from_group(&native.api, m).map(Arc::new))
                .collect()
        })
        .await
    }

    /// Look up a model by its alias.
    pub async fn get_model(&self, alias: &str) -> Result<Arc<Model>> {
        if alias.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Model alias must be a non-empty string".into(),
            });
        }
        let native = self.native.clone();
        let alias = alias.to_owned();
        spawn_blocking(move || match native.get_model(&alias)? {
            Some(m) => Model::from_group(&native.api, m).map(Arc::new),
            None => {
                let available: Vec<String> = native
                    .get_models()
                    .ok()
                    .map(|models| {
                        models
                            .iter()
                            .filter_map(|m| {
                                m.info_ptr().ok().map(|info| unsafe {
                                    crate::detail::api::cstr_to_string((native
                                        .api
                                        .model_api()
                                        .Info_GetAlias)(
                                        info
                                    ))
                                    .unwrap_or_default()
                                })
                            })
                            .collect()
                    })
                    .unwrap_or_default();
                Err(FoundryLocalError::ModelOperation {
                    reason: format!("Unknown model alias '{alias}'. Available: {available:?}"),
                })
            }
        })
        .await
    }

    /// Look up a specific model variant by its unique id.
    ///
    /// NOTE: This will return a `Model` representing a single variant. Use
    /// [`get_model`](Catalog::get_model) to obtain a `Model` with all
    /// available variants.
    pub async fn get_model_variant(&self, id: &str) -> Result<Arc<Model>> {
        if id.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Variant id must be a non-empty string".into(),
            });
        }
        let native = self.native.clone();
        let id = id.to_owned();
        spawn_blocking(move || match native.get_model_variant(&id)? {
            Some(m) => Model::from_variant(&native.api, m).map(Arc::new),
            None => Err(FoundryLocalError::ModelOperation {
                reason: format!("Unknown variant id '{id}'."),
            }),
        })
        .await
    }

    /// Return only the model variants that are currently cached on disk.
    pub async fn get_cached_models(&self) -> Result<Vec<Arc<Model>>> {
        let native = self.native.clone();
        spawn_blocking(move || {
            native
                .get_cached_models()?
                .into_iter()
                .map(|m| Model::from_variant(&native.api, m).map(Arc::new))
                .collect()
        })
        .await
    }

    /// Return model variants that are currently loaded into memory.
    pub async fn get_loaded_models(&self) -> Result<Vec<Arc<Model>>> {
        let native = self.native.clone();
        spawn_blocking(move || {
            native
                .get_loaded_models()?
                .into_iter()
                .map(|m| Model::from_variant(&native.api, m).map(Arc::new))
                .collect()
        })
        .await
    }

    /// Return catalog versions for an alias, optionally narrowed to one model name.
    ///
    /// `max_versions` limits the results per model name; `0` returns all versions.
    pub async fn get_model_versions(
        &self,
        model_alias: &str,
        model_name: Option<&str>,
        max_versions: u32,
    ) -> Result<Vec<Arc<Model>>> {
        if model_alias.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Model alias must be a non-empty string".into(),
            });
        }
        let max_versions =
            i32::try_from(max_versions).map_err(|_| FoundryLocalError::Validation {
                reason: "max_versions must not exceed i32::MAX".into(),
            })?;
        let native = self.native.clone();
        let model_alias = model_alias.to_owned();
        let model_name = model_name.map(str::to_owned);
        spawn_blocking(move || {
            native
                .get_model_versions(&model_alias, model_name.as_deref(), max_versions)?
                .into_iter()
                .map(|m| Model::from_variant(&native.api, m).map(Arc::new))
                .collect()
        })
        .await
    }

    /// Resolve the latest catalog version for the provided model or variant.
    pub async fn get_latest_version(&self, model_or_model_variant: &Model) -> Result<Arc<Model>> {
        let native = self.native.clone();
        let target = model_or_model_variant.selected_native().clone();
        spawn_blocking(move || {
            let latest = native.get_latest_version(&target)?;
            Model::from_variant(&native.api, latest).map(Arc::new)
        })
        .await
    }

    /// Register existing model assets in the local catalog without transferring ownership of those assets.
    pub async fn register_model(
        &self,
        model_path: &str,
        model_id: &str,
        metadata: ModelInfoBuilder,
    ) -> Result<Arc<Model>> {
        if self.catalog_type != CatalogType::Local {
            return Err(FoundryLocalError::Validation {
                reason: "Models can only be registered in the local catalog".into(),
            });
        }
        if model_path.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Model path must be a non-empty string".into(),
            });
        }
        if model_id.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Model id must be a non-empty string".into(),
            });
        }

        let native = self.native.clone();
        let model_path = model_path.to_owned();
        let model_id = model_id.to_owned();
        spawn_blocking(move || {
            let model = native.register_model(&model_path, &model_id, metadata.as_ptr())?;
            Model::from_variant(&native.api, model).map(Arc::new)
        })
        .await
    }

    /// Remove a local registration by alias or model id without deleting the caller-owned model assets.
    pub async fn unregister_model(&self, alias_or_model_id: &str) -> Result<()> {
        if self.catalog_type != CatalogType::Local {
            return Err(FoundryLocalError::Validation {
                reason: "Models can only be unregistered from the local catalog".into(),
            });
        }
        if alias_or_model_id.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "Alias or model id must be a non-empty string".into(),
            });
        }

        let native = self.native.clone();
        let alias_or_model_id = alias_or_model_id.to_owned();
        spawn_blocking(move || native.unregister_model(&alias_or_model_id)).await
    }
}
